consensus/ibft/snapshot_test.go

```diff
@@ -96,32 +109,254 @@ func (ap *testerAccountPool) ValidatorSet() ValidatorSet {
+	return mockVote{validator: validator}
+}
+
+func vote(validator, voted string, auth bool) mockVote {
+	return mockVote{
+		validator: validator,
+		voted:     voted,
+		auth:      auth,
 	}
-	mine := func(validator string) mockVote {
-		return mockVote{validator: validator}
+}
+
+type mockSnapshot struct {
+	validators []string
+	votes      []mockVote
+}
+
+type mockHeader struct {
+	action   mockVote
+	snapshot *mockSnapshot
+}
+
+func newMockHeader(validators []string, vote mockVote) mockHeader {
+	return mockHeader{
+		action: vote,
+		snapshot: &mockSnapshot{
+			validators: validators,
+			votes:      []mockVote{},
+		},
 	}
-	vote := func(validator, voted string, auth bool) mockVote {
-		return mockVote{
-			validator: validator,
-			voted:     voted,
-			auth:      auth,
+}
+
+func buildHeaders(pool *testerAccountPool, genesis *chain.Genesis, mockHeaders []mockHeader) []*types.Header {
+	headers := make([]*types.Header, 0, len(mockHeaders))
+	parentHash := genesis.Hash()
+	for num, header := range mockHeaders {
+		v := header.action
+		pool.add(v.validator)
+
+		h := &types.Header{
+			Number:     uint64(num + 1),
+			ParentHash: parentHash,
+			Miner:      types.ZeroAddress,
+			MixHash:    IstanbulDigest,
+			ExtraData:  genesis.ExtraData,
+		}
+		if v.voted != "" {
+			// if voted is empty, we are just creating a new block
+			// without votes
+			pool.add(v.voted)
+			h.Miner = pool.get(v.voted).Address()
 		}
+		if v.auth {
+			// add auth to the vote
+			h.Nonce = nonceAuthVote
+		} else {
+			h.Nonce = nonceDropVote
+		}
+
+		// sign the vote
+		h = pool.get(v.validator).sign(h)
+		h.ComputeHash()
+
+		parentHash = h.Hash
+		headers = append(headers, h)
 	}
+	return headers
+}
 
-	type mockSnapshot struct {
-		validators []string
-		votes      []mockVote
+func updateHashesInSnapshots(t *testing.T, b *blockchain.Blockchain, snapshots []*Snapshot) {
+	t.Helper()
+	for _, s := range snapshots {
+		hash := b.GetHashByNumber(s.Number)
+		assert.NotNil(t, hash)
+		s.Hash = hash.String()
+	}
+}
+
+func saveSnapshots(t *testing.T, path string, snapshots []*Snapshot) {
+	if snapshots == nil {
+		return
+	}
+
+	store := newSnapshotStore()
+	for _, snap := range snapshots {
+		store.add(snap)
+	}
+	err := store.saveToPath(path)
+	assert.NoError(t, err)
+}
+
+func TestSnapshot_setupSnapshot(t *testing.T) {
+	// Current validators
+	validators := []string{"A", "B", "C", "D"}
+	// New voted validators
+	candidateValidators := []string{"E", "F"}
+
+	pool := newTesterAccountPool()
+	pool.add(validators...)
+	validatorSet := pool.ValidatorSet()
+	genesis := pool.genesis()
+
+	pool.add(candidateValidators...)
+
+	newSnapshot := func(n uint64, set ValidatorSet, votes []*Vote) *Snapshot {
+		return &Snapshot{
+			Number: n,
+			Set:    set,
+			Votes:  votes,
+		}
+	}
+
+	type snapshotData struct {
+		LastBlock uint64
+		Snapshots []*Snapshot
+	}
+	var cases = []struct {
+		name           string
+		epoch          uint64
```

Comment:
Can you please rename this to `epochSize`?
My mind kept thinking it was a specific epoch 🙂

Reply:
I renamed it. Thank you :)